src/context/node.js

```diff
@@ -0,0 +1,354 @@
+   * affected nodes.
+   *
+   * See `Element.assignedSlot` API.
+   *
+   * @param {!Node} node The target node.
+   * @param {!Node} slot The slot to which the target node is assigned.
+   */
+  static assignSlot(node, slot) {
+    if (node[ASSIGNED_SLOT_PROP] == slot) {
+      return;
+    }
+    node[ASSIGNED_SLOT_PROP] = slot;
+    forEachContained(node, (cn) => cn.discover());
+  }
+
+  /**
+   * Unassigns the direct slot previously done by the `assignSlot` call.
+   * Automatically starts the discovery phase for the affected nodes.
+   *
+   * @param {!Node} node The target node.
+   * @param {!Node} slot The slot from which the target node is assigned.
+   */
+  static unassignSlot(node, slot) {
+    if (node[ASSIGNED_SLOT_PROP] != slot) {
+      return;
+    }
+    node[ASSIGNED_SLOT_PROP] = undefined;
+    forEachContained(node, (cn) => cn.discover());
+  }
+
+  /**
+   * Creates the context node and automatically starts the discovery process.
+   *
+   * @param {!Node} node
+   */
+  constructor(node) {
+    /** @const {!Node} */
+    this.node = node;
+
+    /**
+     * Whether this node is a root. The Document DOM nodes are automatically
+     * considered as roots. But other nodes can become roots as well
+     * (e.g. shadow roots) via `setIsRoot()` API.
+     *
+     * @package {boolean}
+     */
+    this.isRoot = node.nodeType == DOCUMENT_NODE;
+
+    /**
+     * The root context node. Always available for a DOM node connected to a
+     * root node after the discovery phase.
+     *
+     * @package {?ContextNode}
+     */
+    this.root = this.isRoot ? this : null;
+
+    /**
+     * Parent should be mostly meaningless to most API clients, because
+     * it's an async concept: a parent context node can can be insantiated at
+     * any time and it doesn't mean that this node has to change. This is
+     * why the API is declared as package-private. However, it needs to be
+     * unobfuscated to avoid cross-binary issues.
+     *
+     * @package {?ContextNode}
+     */
+    this.parent = null;
+
+    /**
+     * See `parent` description.
+     *
+     * @package {?Array<!ContextNode>}
+     */
+    this.children = null;
+
+    /** @private {boolean} */
+    this.parentOverriden_ = false;
+
+    /** @const @private {function()} */
+    this.scheduleDiscover_ = throttleTail(
+      this.discover_.bind(this),
+      setTimeout
+    );
+
+    this.discover();
+  }
+
+  /**
+   * Requests the discovery phase. Asynchronously finds the nearest parent for
+   * this node and its root. Roots and parents set directly via `setParent()`
+   * API are not discoverable.
+   */
+  discover() {
+    if (this.isDiscoverable()) {
+      this.scheduleDiscover_();
+    }
+  }
+
+  /**
+   * Sets (or unsets) the direct parent. If the parent is set, the node will no
+   * longer try to discover itself.
+   *
+   * @param {!ContextNode|!Node|null} parent
+   */
+  setParent(parent) {
+    const parentContext = !parent
+      ? null
```

Comment:
Can this be `parent && parent.nodeType ? ContextNode.get(parent) : parent` ?

Reply:
Yep. That looks better. Done.